src/Utilities/CssDocument.php

```diff
@@ -82,23 +61,47 @@ public function __construct(string $css)
      */
     public function getStyleRulesData(array $allowedMediaTypes): array
     {
-        $splitCss = $this->splitCssAndMediaQuery($allowedMediaTypes);
-
         $ruleMatches = [];
-        foreach ($splitCss as $cssPart) {
-            // process each part for selectors and definitions
-            \preg_match_all('/(?:^|[\\s^{}]*)([^{]+){([^}]*)}/mi', $cssPart['css'], $matches, PREG_SET_ORDER);
-
-            foreach ($matches as $cssRule) {
+        /** @var CssRenderable $rule */
+        foreach ($this->sabberwormCssDocument->getContents() as $rule) {
+            if ($rule instanceof CssAtRuleBlockList && $rule->atRuleName() === 'media') {
+                /** @var string $mediaQueryList */
+                $mediaQueryList = $rule->atRuleArgs();
+                [$mediaType] = \explode('(', $mediaQueryList, 2);
+                if (\trim($mediaType) !== '') {
+                    $mediaTypesExpression = \implode('|', $allowedMediaTypes);
+                    if (!\preg_match('/^\\s*+(?:only\\s++)?+(?:' . $mediaTypesExpression . ')/i', $mediaType)) {
+                        continue;
+                    }
+                }
+                $media = '@media ' . $mediaQueryList;
+                /** @var CssRenderable $nestedRule */
+                foreach ($rule->getContents() as $nestedRule) {
+                    if ($nestedRule instanceof CssDeclarationBlock) {
+                        $ruleMatches[] = [
+                            'media' => $media,
+                            'rule' => $nestedRule,
+                        ];
+                    }
+                }
+            } elseif ($rule instanceof CssDeclarationBlock) {
                 $ruleMatches[] = [
-                    'media' => $cssPart['media'],
-                    'selectors' => $cssRule[1],
-                    'declarations' => $cssRule[2],
+                    'media' => '',
+                    'rule' => $rule,
                 ];
             }
         }
 
-        return $ruleMatches;
+        return \array_map(
+            function (array $ruleMatch): array {
```

Comment:
Can we make the method static?

And can we provide it with more specific type annotations?

Reply:
Done.